From b6164b23d1e1cc09affad1315271bb7a055c253d Mon Sep 17 00:00:00 2001
From: Limeng <Meng.Li@windriver.com>
Date: Wed, 16 Mar 2016 14:59:20 +0800
Subject: [PATCH 7/8] driver: pci: pci-imx6: reset pcie core when link up fail

If there is no PCIE card inserted into PCIE interface, link up will fail.
In this case, original code will disable clock and regulator of pcie
controller and return -EINVAL from imx6_pcie_start_link() function,
and also cause probe exit exceptionally. Now, if you run kexec, the
second kernel will hang up because enters wrong case and access
pcie controller register without enabling pcie controller clock and
regulator.
So, call imx6_pcie_assert_core_reset() function to reset pcie core
when link up fail, so that the second kernel enters correct case when
initializing pcie controller.

Signed-off-by: Meng Li <Meng.Li@windriver.com>
[Quanyang: some modification at context]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/pci/controller/dwc/pci-imx6.c |    1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/pci/controller/dwc/pci-imx6.c b/drivers/pci/controller/dwc/pci-imx6.c
index 1632f2d..aabf7c8 100644
--- a/drivers/pci/controller/dwc/pci-imx6.c
+++ b/drivers/pci/controller/dwc/pci-imx6.c
@@ -1636,6 +1636,7 @@ static int imx_pcie_establish_link(struct imx_pcie *imx_pcie)
 	imx_pcie_reset_phy(imx_pcie);
 
 	if (!IS_ENABLED(CONFIG_PCI_IMX6_COMPLIANCE_TEST)) {
+		imx_pcie_assert_core_reset(imx_pcie);
 		pci_imx_clk_disable(dev);
 		pm_runtime_put_sync(pci->dev);
 		imx_pcie_phy_pwr_dn(imx_pcie);
-- 
1.7.9.5

